add godot-cpp

fmorgner opened the pull request:
Specify library name and version:  **godot-cpp/3.2**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

fmorgner commented:
Aha, i see I can't add two libraries at the same time. Would it be acceptable to add godot_headers separately first and then redo this PR to only include godot-cpp?

Croydon commented:
yes, please split it into two PRs

jgsogo commented:
Here we need to take into account changes in `master` branch... I would love to see Godot here 
![image](https://user-images.githubusercontent.com/1406456/94680895-cdb69b00-0322-11eb-8d89-45c80b3ebe0c.png)

Croydon commented:
> Here we need to take into account changes in master branch... I would love to see Godot here

Same. Getting some Godot tooling here is already great. I would also love to see Godot itself here, but this isn't completely straightforward as they patch many third-party dependencies :/ 

Anyhow, let's not get off-topic here

@fmorgner godot_headers got merged via https://github.com/conan-io/conan-center-index/pull/2980, so could you please update this PR to only include godot-cpp?

fmorgner commented:
AFAICT the `-Og` was only available starting with clang 4.0. Should we patch the `SConstruct` or is there a way to limit the minimum required compiler version for clang in the build?

jgsogo commented:
> AFAICT the `-Og` was only available starting with clang 4.0. Should we patch the `SConstruct` or is there a way to limit the minimum required compiler version for clang in the build?

There are many recipes that raise a `ConanInvalidConfiguration` from the `configure()` method based on different conditions. That's exactly the purpose of that exception, to notify the consumer that the provided configuration is not working. Alternatively, you can also patch the build-system library. Both alternatives are ok.

fmorgner commented:
I have added the compiler version checks (I also realized that `godot-cpp` requires C++14, so I added the standard version check as well) by adapting the implementation used in other recipes. I also bumped the minimum required version of cmake in test package in accordance with https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#kb-h048-cmake-version-required

jgsogo commented on recipes/godot-cpp/all/conanfile.py:
```diff
@@ -0,0 +1,141 @@
+        "os": ["Windows", "Linux", "Macos"],
+        "compiler": ["gcc", "clang", "Visual Studio"],
+        "build_type": ["Debug", "Release", "MinSizeRel", "RelWithDebInfo"],
+    }
```
There is no need to restrict settings (unless we really know it is not working for other combinations). IMHO it is better to leave it free, maybe this same recipe works using other compilers like `intel` or `apple-clang` (probably a bug here).

fmorgner replied:
Fair point. IIRC I would need to set the value of `compiler` to `None`, is that correct?

fmorgner replied:
I was too focused on compilers. I will remove the other limitations as soon as i get home

jgsogo replied:
I would use just:

```
settings = "os", "arch", "compiler", "build_type"
```

It tells Conan that with each different combination of those values it should generate a different package-ID because the binaries generated are different. No need to list explicitly the values, they are stored (and configured) in the `<cache>/settings.yml` file inside the user folder.

jgsogo commented on recipes/godot-cpp/all/conanfile.py:
```diff
@@ -0,0 +1,141 @@
+        "compiler": ["gcc", "clang", "Visual Studio"],
+        "build_type": ["Debug", "Release", "MinSizeRel", "RelWithDebInfo"],
+    }
+    build_requires = ["scons/[~=3]"]
```
We prefer to use fixed versions, also for build-requires: `scons/3.1.2`

jgsogo commented on recipes/godot-cpp/all/conanfile.py:
```diff
@@ -0,0 +1,141 @@
+
+    def imports(self):
+        headers_path = os.path.join(self._source_subfolder, "godot_headers")
+        self.copy("api.json", dst=headers_path, src="res", root_package="godot_headers")
```
Do we need to copy this? Is it not found inside the `headers_dir` directory together with all the `godot_headers`? If not, would it be better to use:

> To use an alternative `api.json` file, add `use_custom_api_file=yes custom_api_file=../api.json`. Be sure to specify the correct location where you placed your file (it can be a relative or absolute path).

fmorgner replied:
Sadly the path to the API file is not dependent on the `headers_dir` (probably a bug upstream). However, i agree it would be better to use the `custom_api_file` argument.

prince-chrismc commented on recipes/godot-cpp/all/conandata.yml:
```diff
@@ -0,0 +1,4 @@
+sources:
+  "cci.3.2-20200130":
```
The version is very unique :worried:

jgsogo replied:
It matches the one from `godot_headers`... I agree it is a bit unique, maybe we can figure out something better in the future.
